client/js/widgets/widget.js

```diff
@@ -1184,22 +1197,99 @@ export class Widget extends StateManaged {
-            jeLoggingRoutineOperationSummary(`${phrase} to ${a.value}`)
+            jeLoggingRoutineOperationSummary(`${phrase} to ${a.value}`);
           else if(a.mode == 'inc' || a.mode == 'dec')
-            jeLoggingRoutineOperationSummary(`${phrase} by ${a.value}`)
+            jeLoggingRoutineOperationSummary(`${phrase} by ${a.value}`);
           else
-            jeLoggingRoutineOperationSummary(`${a.mode} ${phrase}`)
+            jeLoggingRoutineOperationSummary(`${a.mode} ${phrase}`);
         }
       }
 
+      if(a.func == 'TURN') {
+        setDefaults(a, { turn: 1, turnCycle: 'forward', source: 'all', collection: 'TURN' });
+        if([ 'forward', 'backward', 'random', 'position' ].indexOf(a.turnCycle) == -1) {
+          problems.push(`Warning: turnCycle ${a.turnCycle} interpreted as forward.`);
+          a.turnCycle = 'forward'
+        }
+        //copied from select
+        let c = (a.source == 'all' ? Array.from(widgets.values()) : collections[a.source]).filter(w=>w.get('type')=='seat');
+
+        //this get the list of valid index
+        const indexList = []
+        let turn = 1
+        for(const w of c) {
+          if(indexList.indexOf(w.get('index')) == -1 && w.get('player'))
+            indexList.push(w.get('index'));
+          if(w.get('turn'))
+            turn = w.get('index')
+        }
+
+        //loop so it goes for the n next valid index
+        for(let i = 0; i < Math.abs(a.turn) % indexList.length; ++i) {
+          //this checks the next valid index
+          if((a.turnCycle == 'forward' && a.turn > 0) || (a.turnCycle == 'backward' && a.turn < 0)) {
+            indexList.sort((a,b)=>a-b);
+            if(turn >= indexList[indexList.length-1]) {
+              turn = indexList[0];
+            } else {
+              for(const idx of indexList){
+                if(idx > turn) {
+                  turn = idx;
+                  break;
+                }
+              }
+            }
+          } else if((a.turnCycle == 'forward' && a.turn < 0) || (a.turnCycle == 'backward' && a.turn > 0)) {
+            //this checks the previous valid index
+            indexList.sort((a,b)=>b-a);
+            if(turn <= indexList[indexList.length-1]) {
+              turn = indexList[0];
+            } else {
+              for(const idx of indexList) {
+                if(idx < turn) {
+                  turn = idx;
+                  break;
+                }
+              }
+            }
+          }
+        }
+
+        if(a.turnCycle == 'position') {
+          indexList.sort((a,b)=>a-b);
+          if(a.turn == 'first') {
+            turn = indexList[0];
+          } else if(a.turn == 'last') {
+            turn = indexList[indexList.length];
+          } else if(a.turn < 1) {
+            turn = indexList[indexList.length - ((Math.abs(a.turn - 1) % indexList.length) || indexList.length)];
+          } else {
+            turn = indexList[(a.turn - 1) % indexList.length] || 0;
+          }
```

Comment:
In this `if` statement:
1) In the second branch, shouldn't it be `indexList[indexList.length-1]`?
2) In the third branch, if `a.turn == 1`, you will again get `indexList[indexList.length]`.
3) In the fourth branch, is `|| 0` necessary?
4) In the third and fourth branches, what happens if `a.turn` is nonnumeric? (It should probably produce an error, but shouldn't crash the client).